### ISSUE TITLE
`try { let $foo = $foo.bar }` vs `try { let $foo .= bar }` if `.bar` returns an undefined value.

### ISSUE THREAD

**Issue by raiph** [rakudo/rakudo#4486]
## The Problem

There's a consequential inconsistency in handling of `let` statements inside a `try` block.

```
my $foo is default(Nil);
$foo = 'bar'; try   let $foo  =        Nil;      say $foo; # Nil
$foo = 'bar'; try { let $foo .=      &{ Nil } }; say $foo; # Nil
$foo = 'bar'; try { let $foo  = $foo.&{ Nil } }; say $foo; # bar
```
There is at least one issue, and potentially two here:

* `try let ...;` and `try { let ... }` are behaving differently when the value being assigned is `Nil`. That's plausibly entirely appropriate. I'm not sure about `let`'s logic.

* Within a `try` block, `$foo .= bar` and `$foo = $foo.bar` are behaving differently when `.bar` yields an undefined value (including `Nil`). (Note that while I used `.&{...}`, the same problem arises if it's just a regular named method.)

## ~~Expected~~ Desired Behavior

One thing I'd definitely expect is that `try { let $foo = $foo.bar }` and `try { let $foo .= bar }` would behave the same way.

As to whether `try { let $foo = $foo.bar` } should do what `try { let $foo .= bar }` currently does, or vice-versa, I'm torn.

`Nil` is ambiguous. It's a benign failure *or* absence of value. How should `let` respond to an attempt to assign `Nil`?

I currently have no strong view about what it *realistically* needs to do given current roast, and/or expectations of code in the wild, and/or folks' opinions. I've only just discovered this and am mulling what would be the right thing to do.

But right now I'm currently *hoping* `let` will treat assignment of `Nil` as a (benign) failure if it's in a `try` block context.

## More examples focused on `Nil`

my $foo is default(Nil);

$foo = 'bar';           $foo  =        Nil;      say $foo; # Nil
$foo = 'bar';       let $foo  =        Nil;      say $foo; # Nil
$foo = 'bar'; try   let $foo  =        Nil;      say $foo; # Nil
$foo = 'bar'; try { let $foo  =        Nil };    say $foo; # bar

$foo = 'bar';           $foo  =        Nil;      say $foo; # Nil
$foo = 'bar';       let $foo  =        Nil;      say $foo; # Nil
$foo = 'bar'; try   let $foo  =        Nil;      say $foo; # Nil
$foo = 'bar'; try { let $foo  =        Nil };    say $foo; # bar

$foo = 'bar';           $foo  = $foo.&{Nil};     say $foo; # Nil
$foo = 'bar';       let $foo  = $foo.&{Nil};     say $foo; # Nil
$foo = 'bar'; try   let $foo  = $foo.&{Nil};     say $foo; # Nil
$foo = 'bar'; try { let $foo  = $foo.&{ Nil } }; say $foo; # bar

$foo = 'bar';           $foo .=      &{ Nil };   say $foo; # Nil
$foo = 'bar';       let $foo .=      &{ Nil };   say $foo; # Nil
$foo = 'bar'; try   let $foo .=      &{ Nil };   say $foo; # Nil
$foo = 'bar'; try { let $foo .=      &{ Nil } }; say $foo; # Nil

**Comment by vrurg** [comment]
I have some preliminary results on this issue. Not sure if you is going to like them, but...

First of all, _Nil_ has nothing to do with the observed behavior here. It served you as a good helper and nothing else aside of this. The rest is well-documented. Let's start with the fact that [let](https://docs.raku.org/language/operators#prefix_let) (and, BTW, `temp` for that same matter) is an operator. `.=` is too. But `let` has [looser precedence](https://docs.raku.org/language/operators#Operator_precedence). So in

```raku
let $foo .=      &{ Nil }
```

when `let` is invoked `$foo` _already_ has the result of `$foo.&{ Nil }` in it. And this is what you get when the block ends returning `Nil`. There is immediate workaround, pretty much evident though:

```raku
(let $foo) .= &{ Nil };
```

And it works as expected.

There is also a detail you possibly overlooked: `let` is not only about exception throwing. Or rather this is a side effect. In first place it is about blocks returning undefined values. In other words, success of a block is measured by definedness of the return value. It is based on `UNDO` phaser, but it's documentation is rather succinct and doesn't clearly specify what 'unsuccessful block exit' means. Anyway, all of your `try` blocks in the provided examples are simply redundant because no code under them throws. Those doing what you want them to do succeed because of the assignments returning Nil. Well, as a matter of fact, those not doing what is expected also do the same thing, as I explained above. :) 

And another `try`-related note is about `try let $foo` cases. Raku doesn't create a scope for them because there is no explicit block declaration. So, the statement is simply wrapped into `try`, but it belongs to the same scope where `my $foo` is declared. This basically draws `let` useless because even if both `my` and `try` reside in a block, `$foo` will be dropped as soon as the block ends and the only visible effect you'd get is certain slowdown on block termination as it might be invoking `UNDO` phaser installed for `let`. Apparently, it might not run the phaser either and it certainly will not depend on the outcome of your `try let $foo ...` statement unless it is the last one.

However disappointing the above written looks, I may have some hope for you here. I tried playing with operator priorities by giving `let` the same level as `.` and `.=` have. This makes the `(let $foo) .= method` order default and results in expected outcome of `$foo` falling back to _"bar"_. Moreover, the change didn't break specs. Unfortunately, I'm not ready to call this a fix as there might be some side effects I'm not aware of yet. And some ecosystem modules may not work with the change.

**Comment by raiph** [comment]
First, I think it's time to close this issue, right?

----

Feel free to skip reading the rest of this comment, and certainly replying to it in any way beyond a short note, but the following might be of interest to you and/or help emphasize how much I appreciate your awesome follow up. :)

> Not sure if you are going to like [preliminary results], but...

I *love* them because what I had hoped for was substantive insight into the problem, and possible solutions, and I very much now have that. I see you've also followed up with a WAT issue too; I'll be reading that after replying to this. Thank you for such a robust response!

> First of all, `Nil` has nothing to do with the observed behavior here. It served you as a good helper and nothing else aside of this. 

Right. Fwiw I knew that -- it's why I changed the title and made some other edits to my comment. But moving on...

> The rest is well-documented. Let's start with the fact that `let` (and, BTW, `temp` for that same matter) is an operator. `.=` is too. But `let` has looser precedence.

This is the key. I even knew this at one time! I had just forgotten, and made the fatal mistake of not rereading the doc. Now it all instantly makes sense. Thanks for your patience and I will admonish myself for not rereading the doc and thinking things through enough. And/or contribute to the WAT issue you've filed.

> So in `let $foo .=      &{ Nil }` when `let` is invoked `$foo` already has the result of `$foo.&{ Nil }` in it. And this is what you get when the block ends returning `Nil`. There is immediate workaround, pretty much evident though: `(let $foo) .= &{ Nil };` And it works as expected.

Totally makes (technical) sense.

> There is also a detail you possibly overlooked: `let` is not only about exception throwing. Or rather this is a side effect. In first place it is about blocks returning undefined values.

Fwiw I knew that.

> all of your `try` blocks in the provided examples are simply redundant because no code under them throws.

Right. In the original scenario that led to me opening this issue the `try` had its merit (in case the `.bar` method mentioned in the title threw) but in writing the examples I used `.&{Nil}` to try keep things simple.

> And another `try`-related note is about `try let $foo` cases. ... This basically draws `let` useless

If that's so, then perhaps an warning would be warranted? But I'll defer further comment for the WAT issue.

> However disappointing the above written looks, I may have some hope for you here. I tried playing with operator priorities by giving `let` the same level as `.` and `.=` have. This makes the `(let $foo) .= method` order default and results in expected outcome of `$foo` falling back to `"bar"`. Moreover, the change didn't break specs.

Interesting.

> Unfortunately, I'm not ready to call this a fix as there might be some side effects I'm not aware of yet. And some ecosystem modules may not work with the change.

Right. This was what I meant by "I currently have no strong view about what it realistically needs to do given current roast, and/or expectations of code in the wild, and/or folks' opinions. I've only just discovered this and am mulling what would be the right thing to do."

And that was *before* I learned I had the wrong end of the stick anyway. :)

**Comment by vrurg** [comment]
> First, I think it's time to close this issue, right?

Up to you. Can be kept until the problem-solving is resolved and PR is merged. Though I'm anyway inclined to merge it.

> > And another `try`-related note is about `try let $foo` cases. ... This basically draws `let` useless
> 
> If that's so, then perhaps an warning would be warranted?

I don't think so. First of all, this could be done on purpose. Not sure if `quiet try let $foo = ...` appeals to my aesthetic feelings.  Second, understanding the difference between a block and a statement and how `<keyword> <statement>` is different from `<keyword> { <statments> }` is something better taken care of by books and documentation, not by the compiler. Or we'd have to issue a warning even for `try my $foo = 42;`.

**Comment by raiph** [comment]
>>> `try let $foo` cases. ... basically draws `let` useless

>> If that's so, then perhaps an warning would be warranted?

> this could be done on purpose.

I think I've misinterpreted your words "cases" and "useless". If you just meant my `let $foo = Nil` cases, then sure, they're useless (except as examples to make the point that the very last example was different from the others, which was the only reason I wrote them). But if you meant that `try let $foo` *in general* was useless, then I'd presume a warning would be helpful; if someone does something useless *on purpose*, it's surely near guaranteed that they don't realize it's useless. 

> Or we'd have to issue a warning even for `try my $foo = 42;`

Unless it's because of me misunderstanding what you meant by your prior declaration of things being useless, I'm not following your logic. How is that useless *in the general case*? Consider `try my $foo = bar;`. How is it useless to use `try`?

**Comment by vrurg** [comment]
I certainly wasn't clear. It sometimes hard to be when the picture is clear in my mind. :)

`try let $foo` is only useless in your example because it is used in the global context. BTW, that's where a warning might make sense, actually. Otherwise, of course

```raku
my $foo = 42;
{
    try let $foo = bar;
}
```

do what's expected.

**Comment by vrurg** [comment]
Resolved with #4492
